File: nand2tetris/jack-compiler/scala/src/main/scala/Ast.scala
package jackcompiler.ast

import jackcompiler.Token
import jackcompiler.Kind


// ------------------- Inicio da AST

abstract class Node {
    def accept(v: Visitor) :Unit
}

abstract class Expression extends Node 
abstract class Statement extends Node
abstract class Identifier extends Expression


case class VarDeclaration (val kind :Kind.Kind, val varType :String, val name :String  ) extends Node {
     def accept (v: Visitor) = {
            return v.visitVarDeclaration(this)
    }
}

case class ClassDec (name: String,  classVardecs :List[VarDeclaration], subroutineDecs :List[Subroutine]) extends Node {
    def accept (v: Visitor) = {
        return v.visitClassDec(this)

    }
}



case class Subroutine (modifier:String, funcType:String, name: String, params:List[VarDeclaration],vars:List[VarDeclaration], statements: Statements) extends Node {
    def accept (v: Visitor) = {
        return v.visitSubroutine(this)
    }
}





case class BinaryExpression (val left:Expression, val operator: Char, val right: Expression ) extends jackcompiler.ast.Expression {
    override def accept (v: Visitor) = {
        return v.visitBinaryExpression(this)
    }
}


case class UnaryExpression (val operator: Char, val right: Expression ) extends jackcompiler.ast.Expression {
    override def accept (v: Visitor) = {
        return v.visitUnaryExpression(this)
    }
}

case class Variable (val varName:String) extends Identifier  {
    
    override def accept (v: Visitor) = {
        return v.visitVariable(this)
    }
}

case class IndexVariable (val varName: String, exp: Expression) extends Identifier  {
    override def accept (v: Visitor) = {
        return v.visitIndexVariable(this)
    }
}
case class IntegerLiteral (val value:Int) extends Expression  {

    override def accept (v: Visitor) = {
        return v.visitIntegerLiteral(this)
    }

}

case class StringLiteral (val value:String) extends Expression  {

    override def accept (v: Visitor) = {
        return v.visitStringLiteral(this)
    }

}

case class KeywordLiteral (val value:String) extends Expression  {

    override def accept (v: Visitor) = {
        return v.visitKeywordLiteral(this)
    }

}

case class Call (val name:String, arguments:List[Expression]) extends Expression  {

    override def accept (v: Visitor) = {
        return v.visitCall(this)
    }

}

case class Statements (val sts:List[Statement]) extends Node {

    override def accept (v: Visitor) = {
        return v.visitStatements(this)
    }

}

case class LetStatement (val id:Identifier, val exp:Expression) extends Statement{
    def accept (v: Visitor) = {
        return v.visitLetStatement(this)
    }
}


case class IfStatement (val condition:Expression, val thenBranch: Statements, val elseBranch:Option[Statements]=None) extends Statement{
    def accept (v: Visitor) = {
        return v.visitIfStatement(this)
    }
}


case class WhileStatement (val condition:Expression, val body: Statements) extends Statement{
    def accept (v: Visitor) = {
        return v.visitWhileStatement(this)
    }
}

case class ReturnStatement (val value:Option[Expression]=None) extends Statement{
    def accept (v: Visitor) = {
        return v.visitReturnStatement(this)
    }
}


case class DoStatement (val subroutine:Expression) extends Statement{
    def accept (v: Visitor) = {
        return v.visitDoStatement(this)
    }
}

// ------------------- fim da AST

abstract class Visitor {


    def visitClassDec(v:ClassDec) : Unit
    def visitSubroutine(v:Subroutine) : Unit
    

    def visitVarDeclaration(v:VarDeclaration) : Unit


    def visitStatements (v: Statements) : Unit
    def visitLetStatement (v: LetStatement) : Unit
    def visitIfStatement (v: IfStatement) : Unit
    def visitWhileStatement (v: WhileStatement) : Unit
    def visitReturnStatement (v: ReturnStatement) : Unit
    def visitDoStatement (v: DoStatement) : Unit


    def visitVariable (v: Variable) : Unit
    def visitIndexVariable (v: IndexVariable) : Unit
    def visitIntegerLiteral (v: IntegerLiteral) : Unit
    def visitStringLiteral (v: StringLiteral) : Unit
    def visitKeywordLiteral (v: KeywordLiteral) : Unit
    def visitBinaryExpression (v: BinaryExpression) : Unit
    def visitUnaryExpression (v: UnaryExpression) : Unit
    def visitCall (v: Call) : Unit
    
    

    
    
}

File: compiler/ccompiler/ast/ast.go
package ast

import (
	"bytes"

	"ccompiler/token"
)

/*
program = Program(function_declaration)
function_declaration = Function(string, statement) //string is the function name
statement = Return(exp)
exp = Constant(int)
*/

// Node represents an AST node.
type Node interface {
	TokenLiteral() string
	String() string
}

type Statement interface {
	Node
	statementNode()
}

type Expression interface {
	Node
	expressionNode()
}

type Program struct {
	Function *Function
}

// TokenLiteral returns the first token literal of a program.
func (p *Program) TokenLiteral() string {
	return ""
}

func (p *Program) String() string {
	var out bytes.Buffer

	out.WriteString("(Program ")

	out.WriteString(p.Function.String())

	out.WriteString(" )")

	return out.String()
}

type Function struct {
	Token      token.Token
	Statements []Statement
}

func (fl *Function) expressionNode() {}

// TokenLiteral returns a token literal of function.
func (fl *Function) TokenLiteral() string {
	return fl.Token.Literal
}

func (fl *Function) String() string {
	var out bytes.Buffer

	out.WriteString("( " + fl.TokenLiteral() + " ")

	for _, st := range fl.Statements {
		out.WriteString(st.String())
	}

	out.WriteString(" )")

	return out.String()
}

// ReturnStatement represents a return statement.
type ReturnStatement struct {
	Token       token.Token // the token.RETURN token
	ReturnValue Expression
}

func (rs *ReturnStatement) statementNode() {}

// TokenLiteral returns a token literal of return statement.
func (rs *ReturnStatement) TokenLiteral() string {
	return rs.Token.Literal
}

func (rs *ReturnStatement) String() string {
	var out bytes.Buffer

	out.WriteString(rs.TokenLiteral() + "  ")

	if rs.ReturnValue != nil {
		out.WriteString(rs.ReturnValue.String())
	}

	out.WriteString(";")

	return out.String()
}

// IntegerLiteral represents an integer literal.
type IntegerLiteral struct {
	Token token.Token
	Value int64
}

func (il *IntegerLiteral) expressionNode() {}

// TokenLiteral returns a token literal of integer.
func (il *IntegerLiteral) TokenLiteral() string {
	return il.Token.Literal
}

func (il *IntegerLiteral) String() string {
	return il.Token.Literal
}

type UnaryExpression struct {
	Operator token.Token
	Right    Expression
}

func (exp *UnaryExpression) expressionNode()      {}
func (exp *UnaryExpression) TokenLiteral() string { return exp.Operator.Literal }
func (exp *UnaryExpression) String() string {
	return exp.Operator.Literal
}

type VarExpression struct {
	Name token.Token
}

func (exp *VarExpression) expressionNode()      {}
func (exp *VarExpression) TokenLiteral() string { return exp.Name.Literal }

func (exp *VarExpression) String() string {
	var out bytes.Buffer

	out.WriteString(exp.TokenLiteral())

	return out.String()
}

type DeclareStatement struct {
	Type    token.Token
	VarName token.Token
	Init    Expression
}

func (rs *DeclareStatement) statementNode() {}

func (rs *DeclareStatement) TokenLiteral() string {
	return rs.VarName.Literal
}

func (rs *DeclareStatement) String() string {
	var out bytes.Buffer

	out.WriteString("(Decl ")

	out.WriteString(rs.TokenLiteral())

	if rs.Init != nil {
		out.WriteString("=")
		out.WriteString(rs.Init.String())
	}

	out.WriteString(") ")

	return out.String()
}

type BinaryExpression struct {
	Operator token.Token
	Left     Expression
	Right    Expression
}

func (exp *BinaryExpression) expressionNode()      {}
func (exp *BinaryExpression) TokenLiteral() string { return exp.Operator.Literal }

func (exp *BinaryExpression) String() string {
	var out bytes.Buffer

	out.WriteString("(" + exp.Operator.Literal + " ")

	out.WriteString(exp.Left.String() + " ")
	out.WriteString(exp.Right.String() + " ")

	out.WriteString(" )")

	return out.String()
}

type AssignStatement struct {
	VarName token.Token
	Init    Expression
}

func (rs *AssignStatement) statementNode() {}

func (rs *AssignStatement) TokenLiteral() string {
	return rs.VarName.Literal
}

func (rs *AssignStatement) String() string {
	var out bytes.Buffer

	out.WriteString("(Assign ")

	out.WriteString(rs.TokenLiteral())

	if rs.Init != nil {
		out.WriteString("=")
		out.WriteString(rs.Init.String())
	}

	out.WriteString(") ")

	return out.String()
}
